internal/configs/provider.go

```diff
@@ -145,6 +160,109 @@ func decodeProviderBlock(block *hcl.Block) (*Provider, hcl.Diagnostics) {
 	return provider, diags
 }
 
+func (p *Provider) decodeStaticFields(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if p.ForEach != nil {
+		if eval == nil {
+			return nil, diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Iteration not allowed in test files",
+				Detail:   "for_each was declared as an provider attribute in a test file",
+				Subject:  p.AliasExpr.Range().Ptr(),
+			})
+		}
+
+		foeachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
+			var diags tfdiags.Diagnostics
+			evalContext, evalDiags := eval.EvalContext(StaticIdentifier{
+				Module:    eval.call.addr,
+				Subject:   fmt.Sprintf("provider.%s.for_each", p.Name),
+				DeclRange: p.ForEach.Range(),
+			}, refs)
+			return evalContext, diags.Append(evalDiags)
+		}
+
+		forVal, evalDiags := evalchecks.EvaluateForEachExpression(p.ForEach, foeachRefsFunc)
+		diags = append(diags, evalDiags.ToHCL()...)
+		if evalDiags.HasErrors() {
+			return nil, diags
+		}
+
+		var out []*Provider
+		for k, v := range forVal {
+			v := v
+			iter := *p
+			iter.Alias = k
```

Comment:
We probably want to check `hclsyntax.ValidIdentifier(alias)`, not sure how the numeric count will handle that name.0 for example.